### PR TITLE
Tx approval wait modal

### PR DESCRIPTION
**Pull request by Velenir**
Global Modal with an escape hatch of Singleton component + `export let` (you ever seen that one? :wink: )
To use for _tx taking to long_ action
Only basic styling for now

Also, I thought to disable Close `x` button to force the user to click Yes/No, but it removes the whole header with title :cry: 

localhost_8080_trade_USDC-DAI_sell=1 price=0 99722 from= expires=

### CODE REVIEW

**Comment by gnosis-info** [comment]
Travis automatic deployment:
   * **🎩 [Dapp](https://pr1549--dexreact.review.gnosisdev.com)**: Testing web app
   * **📚 [Storybook](https://pr1549--dexreact.review.gnosisdev.com/storybook/)**: Component stories

File: src/api/wallet/composeProvider.ts [hunk 1]
```diff
@@ -73,18 +75,13 @@ const wrapInTimeout = (middleware: JsonRpcMiddleware, timeout = DEFAULT_TX_APPRO
 
     let timeoutId: NodeJS.Timeout | null = null
 
-    timeoutId = setTimeout(function askOnTimeout() {
-      if (confirm('Stop waiting for tx approval?')) {
+    timeoutId = setTimeout(async function askOnTimeout() {
+      if (await openWaitForTxApprovalModal(timeout)) {
```
Comment (anxolin) [inline src/api/wallet/composeProvider.ts:79]: This has the undesired effect that on multiple tx, you get a lot of confirmations

Can we do sth in the direction of keeping a count of the number of transactions waiting.

- When you send transactions to the wallet, you increment
- When they resolve or reject, you decrement
- When the timer fires, we ask for confirmation to stop waiting, and if the user cancels the waiting, we decrement

Only when our count is 0, we can stop waiting.

**Review by anxolin — COMMENTED** [review]
This is a great PR. I have some concern with the modal becoming annoying when you send several transactions.

Please, remember to increment a lot the timer before merging. Is 60s too much?

File: src/api/wallet/composeProvider.ts (hunk 1 above)
Comment (Velenir) [reply src/api/wallet/composeProvider.ts:79]: Thankfully we can't send multiple trade txs, but yes there are cases.
Need a count somewhere

Comment (Velenir) [reply src/api/wallet/composeProvider.ts:79]: Not sure how to go about implementing this
If we use one modal for all pending txs that timed out:
- Do we just not close it if the count > 0
- Do we close and reopen
- Do we allow to stop waiting on all txs together
- Do we update the modal in real time if another tx becomes late
- Maybe throttle all tx approvals, but then they can potentially come out of order, depending on wallet

I don't see it being a good UX either way

@W3stside chime in please

**Comment by Velenir** [comment]
localhost_8080_trade_USDC-DAI_sell=1 price=1 00222 from= expires=

File: src/api/wallet/composeProvider.ts (hunk 1 above)
Comment (W3stside) [reply src/api/wallet/composeProvider.ts]: oof. uhhhh shit. idk what to input here... This sounds super edgecase-y and the amount of work seems... heavy.

But a global state var keeping track of current ongoing txs is a start, then we have 1 modal? and wait there? add any new incoming @Velenir

**Comment by W3stside** [comment]
@Velenir nice change

Maybe change title to be more inline with errors so:
```js
Transaction Alert: No response from wallet for pending transaction
```

File: src/components/OuterModal/SingletonModal.tsx [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  rightButton: Modal.Button,
+}
+
+const defaultModalOptions: OuterModalOptions = {
+  message: null,
+  ...defaultButtons,
+}
+
+export let openGlobalModal: (params: FillAndToggleModal) => Promise<boolean> = () => Promise.resolve(false)
+export let closeGlobalModal: () => void = () => void 0
```
Comment (W3stside) [inline src/components/OuterModal/SingletonModal.tsx:32]: what is happening here loool nooo. new code scary

**Review by W3stside — APPROVED** [review]

File: src/components/OuterModal/WaitForTxApproval.tsx [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  pendingTxApprovals: Set<number | string>
+}
+
+const PenidngTxApprovalsMessage: React.FC<{ txNumber: number }> = ({ txNumber }) => {
+  return (
+    <>
+      <h4>No response from wallet for pending transaction</h4>
+      <p>There are currently {txNumber} transactions waiting for approval for your wallet.</p>
+      <p>
+        Mesa has detected a potential transaction timeout. Please validate with your wallet that you have properly
```
Comment (anxolin) [inline src/components/OuterModal/WaitForTxApproval.tsx:21]: I would drop this phrase completly. `Mesa has detected a potential transaction timeout`

Also, not nice to hardcode "Mesa", Mesa is dxDAO product not ours

File: src/components/OuterModal/WaitForTxApproval.tsx [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    </>
+  )
+}
+
+const NoTxsWillClose: React.FC = () => {
+  return (
+    <>
+      <h4>All transactions have been handled</h4>
+      <p>Mesa has received responses for all pending transactions.</p>
+      <p> This modal will close in a moment.</p>
```
Comment (anxolin) [inline src/components/OuterModal/WaitForTxApproval.tsx:34]: remoe leading space?

File: src/components/OuterModal/WaitForTxApproval.tsx [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+  const [initialHeight, setInitialHeight] = useState(0)
+
+  const wrapperRef = useRef<HTMLDivElement>(null)
+
+  // layout effect for sync rerendering on style change
+  // otherwise we get layout thrashing
+  useLayoutEffect(() => {
+    if (!wrapperRef.current) return
+    setInitialHeight(wrapperRef.current.offsetHeight)
+  }, [])
```
Comment (anxolin) [inline src/components/OuterModal/WaitForTxApproval.tsx:54]: should I ask?

**Review by anxolin — APPROVED** [review]

File: src/components/OuterModal/WaitForTxApproval.tsx (hunk 5 above)
Comment (Velenir) [reply src/components/OuterModal/WaitForTxApproval.tsx:54]: LayoutEffect is triggered synchronously before browser paint
with the usual Effect it'll `render first-> paint-> measure -> change style -> render -> paint`, you'll see a slight jump on style change
with Layout Effect `render first-> measure -> change style -> render -> paint`, we trigger rerender before browser paints

**Pull request merged**